File: Kalorie/Kalorie/Core/UseCases/FetchMyFoodItemReportUseCase.swift
//
//  FetchMyFoodItemReportUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 17.09.2026.
//

import Foundation

protocol FetchMyFoodItemReportUseCaseProtocol {
    func callAsFunction(barcode: String) async throws -> FoodItemReportDomain?
}

struct FetchMyFoodItemReportUseCase: FetchMyFoodItemReportUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(barcode: String) async throws -> FoodItemReportDomain? {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let dto: FoodItemReportDTO? = try await dataProvider.loadAsync(
            id: FoodItemReportDomain.id(barcode: barcode, userId: userId),
            from: Constants.Firestore.foodItemReports
        )
        return dto?.asDomain()
    }
}

#if DEBUG
struct FetchMyFoodItemReportUseCaseFake: FetchMyFoodItemReportUseCaseProtocol {

    // MARK: - Properties

    var stubbedReport: FoodItemReportDomain?
    var errorToThrow: Error?

    // MARK: - Functions

    func callAsFunction(barcode: String) async throws -> FoodItemReportDomain? {
        if let errorToThrow { throw errorToThrow }
        return stubbedReport
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/FetchMyFoodItemReportUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodItemReportDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodItemReportDTO
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.utils.Constants

interface FetchMyFoodItemReportUseCaseProtocol {
    suspend operator fun invoke(barcode: String): FoodItemReportDomain?
}

class FetchMyFoodItemReportUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : FetchMyFoodItemReportUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(barcode: String): FoodItemReportDomain? {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val dto: FoodItemReportDTO? = dataProvider.loadAsync(
            id = FoodItemReportDomain.id(barcode = barcode, userId = userId),
            from = Constants.Firestore.FOOD_ITEM_REPORTS,
        )
        return dto?.asDomain()
    }
}
